File: Core/GrowthWare.BusinessLogic/BGroups.cs
using GrowthWare.DataAccess.Interfaces;
using GrowthWare.Framework;
using GrowthWare.Framework.Models;
using System;
using System.Collections;
using System.Data;

namespace GrowthWare.BusinessLogic
{
    /// <summary>
    /// Process business logic for accounts
    /// </summary>
    /// <remarks>
    /// <![CDATA[
    /// MSecurityEntity can be found in the GrowthWare.Framework.Model.Profiles namespace.  
    /// 
    /// The following properties are necessary for correct business logic operation.
    /// .ConnectionString
    /// .DALName
    /// .DALNameSpace
    /// ]]>
    /// </remarks>
    /// <example> This sample shows how to create an instance of the class.
    /// <code language="VB.NET">
    /// <![CDATA[
    /// Dim myBll as new BGroups(mySecurityEntityProfile)
    /// ]]>
    /// </code>
    /// </example>
    public class BGroups : BaseBusinessLogic
    {
        private IGroups m_DGroups;

        /// <summary>
        /// Private BGroups() to ensure only new instances with passed parameters is used.
        /// </summary>
        /// <remarks></remarks>
        private BGroups()
        {
        }

        /// <summary>
        /// Parameters are need to pass along to the factory for correct connection to the desired datastore.
        /// </summary>
        /// <param name="securityEntityProfile">The Security Entity profile used to obtain the DAL name, DAL name space, and the Connection String</param>
        /// <param name="centralManagement">Boolean value indicating if the system is being used to manage multiple database instances.</param>
        /// <remarks></remarks>
        /// <example> This sample shows how to create an instance of the class.
        /// <code language="VB.NET">
        /// <![CDATA[
        /// MSecurityEntity MSecurityEntity = MSecurityEntity = New MSecurityEntity();
        /// MSecurityEntity.Id = ConfigSettings.DefaultSecurityEntityID;
        /// MSecurityEntity.DAL = ConfigSettings.DAL;
        /// MSecurityEntity.DAL_Namespace = ConfigSettings.DAL_NameSpace(MSecurityEntity.DAL);
        /// MSecurityEntity.DAL_Name = ConfigSettings.DAL_AssemblyName(MSecurityEntity.DAL);
        /// MSecurityEntity.ConnectionString = ConfigSettings.ConnectionString;
        /// 
        /// BGroupss mBGroups = BGroupss = New BGroupss(MSecurityEntity, ConfigSettings.CentralManagement);
        /// ]]>
        /// </code>
        /// <code language="C#">
        /// <![CDATA[
        /// Dim MSecurityEntity As MSecurityEntity = New MSecurityEntity()
        /// MSecurityEntity.Id = ConfigSettings.DefaultSecurityEntityID
        /// MSecurityEntity.DAL = ConfigSettings.DAL
        /// MSecurityEntity.DAL_Namespace = ConfigSettings.DAL_NameSpace(MSecurityEntity.DAL)
        /// MSecurityEntity.DAL_Name = ConfigSettings.DAL_AssemblyName(MSecurityEntity.DAL)
        /// MSecurityEntity.ConnectionString = ConfigSettings.ConnectionString
        /// 
        /// Dim mBGroups As BGroupss = New BGroupss(MSecurityEntity, ConfigSettings.CentralManagement)
        /// ]]>
        /// </code>
        /// </example>
        public BGroups(MSecurityEntity securityEntityProfile, bool centralManagement)
        {
            if (securityEntityProfile == null)
            {
                throw new ArgumentNullException("securityEntityProfile", "The securityEntityProfile cannot be a null reference (Nothing in Visual Basic)!!");
            }
            if (centralManagement)
            {
                if (m_DGroups == null)
                {
                    m_DGroups = (IGroups)ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DGroups");
                }
            }
            else
            {
                m_DGroups = (IGroups)ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DGroups");
            }

            m_DGroups.ConnectionString = securityEntityProfile.ConnectionString;
            m_DGroups.SecurityEntitySeqID = securityEntityProfile.Id;
        }

        /// <summary>
        /// Gets the groups by security entity.
        /// </summary>
        /// <param name="securityEntityId">The security entity ID.</param>
        /// <returns>DataTable.</returns>
        public DataTable GetGroupsBySecurityEntity(int securityEntityId)
        {
            MGroupProfile myProfile = new MGroupProfile();
            myProfile.SecurityEntityId = securityEntityId;
            m_DGroups.Profile = myProfile;
            return m_DGroups.GroupsBySecurityEntity();
        }

        /// <summary>
        /// Adds the group.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns>System.Int32.</returns>
        public void AddGroup(MGroupProfile profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!");
            m_DGroups.Profile = profile;
            if(DatabaseIsOnline()) m_DGroups.AddGroup();
        }

        /// <summary>
        /// Gets the profile.
        /// </summary>
        /// <param name="groupId">The group ID.</param>
        /// <returns>MGroupProfile.</returns>
        public MGroupProfile GetProfile(int groupId)
        {
            MGroupProfile retProfile = new MGroupProfile();
            retProfile.Id = groupId;
            m_DGroups.Profile = retProfile;
            retProfile = new MGroupProfile(m_DGroups.ProfileData());
            return retProfile;
        }

        /// <summary>
        /// Deletes the group.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        public bool DeleteGroup(MGroupProfile profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!");
            bool retVal = false;
            m_DGroups.Profile = profile;
            if (DatabaseIsOnline()) retVal = m_DGroups.DeleteGroup();
            return retVal;
        }

        /// <summary>
        /// Updates the group.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        public void Save(MGroupProfile profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!");
            m_DGroups.Profile = profile;
            if (DatabaseIsOnline()) m_DGroups.Save();
        }

        /// <summary>
        /// Gets the selected roles.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns>System.String[][].</returns>
        public string[] GetSelectedRoles(MGroupRoles profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!");
            ArrayList ClientRoles = new ArrayList();
            m_DGroups.GroupRolesProfile = profile;
            DataTable myDataTable = null;
            if (DatabaseIsOnline()) 
            {
                try
                {
                    myDataTable = m_DGroups.GroupRoles();
                    DataRow myDR = null;
                    foreach (DataRow myDR_loopVariable in myDataTable.Rows)
                    {
                        myDR = myDR_loopVariable;
                        ClientRoles.Add(myDR["Role"].ToString());
                    }
                }
                catch (BusinessLogicLayerException)
                {
                    throw;
                }
            }
            return (string[])ClientRoles.ToArray(typeof(string));
        }

        /// <summary>
        /// Searches groups using the specified search criteria.
        /// </summary>
        /// <param name="searchCriteria">The search criteria.</param>
        /// <returns>DataTable.</returns>
        public DataTable Search( MSearchCriteria searchCriteria)
        {
            if (searchCriteria == null) throw new ArgumentNullException("searchCriteria", "searchCriteria cannot be a null reference (Nothing in Visual Basic)!!");
            DataTable mRetVal = null;
            if (DatabaseIsOnline()) mRetVal = m_DGroups.Search(searchCriteria);
            return mRetVal;
        }

        /// <summary>
        /// Updates the group roles.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        public bool UpdateGroupRoles(MGroupRoles profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!");
            bool mRetVal = false;
            m_DGroups.GroupRolesProfile = profile;
            if (DatabaseIsOnline()) mRetVal = m_DGroups.UpdateGroupRoles();
            return mRetVal;
        }
    }
}

File: Net_Framework/VB/GrowthWare.Framework.BusinessData/BusinessLogicLayer/BGroups.vb
﻿Imports GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces
Imports GrowthWare.Framework.Common
Imports GrowthWare.Framework.Model.Profiles
Imports GrowthWare.Framework.BusinessData.DataAccessLayer

Namespace BusinessLogicLayer
    Public Class BGroups
        Inherits BaseBusinessLogic
        Private m_DGroups As IDGroups

        ''' <summary>
        ''' Private sub new() to ensure only new instances with passed parameters is used.
        ''' </summary>
        ''' <remarks></remarks>
        Private Sub New()

        End Sub

        Public Function Save(ByVal profile As MGroupProfile) As Integer
            m_DGroups.Profile = profile
            Return m_DGroups.Save()
        End Function

        ''' <summary>
        ''' Parameters are need to pass along to the factory for correct connection to the desired datastore.
        ''' </summary>
        ''' <param name="securityEntityProfile">The Security Entity profile used to obtain the DAL name, DAL name space, and the Connection String</param>
        ''' <param name="centralManagement">Boolean value indicating if the system is being used to manage multiple database instances.</param>
        ''' <remarks></remarks>
        ''' <example> This sample shows how to create an instance of the class.
        ''' <code language="VB.NET">
        ''' <![CDATA[
        ''' MSecurityEntityProfile mSecurityEntityProfile = MSecurityEntityProfile = New MSecurityEntityProfile();
        ''' mSecurityEntityProfile.ID = ConfigSettings.DefaultSecurityEntityID;
        ''' mSecurityEntityProfile.DAL = ConfigSettings.DAL;
        ''' mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL);
        ''' mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL);
        ''' mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString;
        ''' 
        ''' BGroups mBGroups = New BGroups(mSecurityEntityProfile, ConfigSettings.CentralManagement);
        ''' ]]>
        ''' </code>
        ''' <code language="C#">
        ''' <![CDATA[
        ''' Dim mSecurityEntityProfile As MSecurityEntityProfile = New MSecurityEntityProfile()
        ''' mSecurityEntityProfile.ID = ConfigSettings.DefaultSecurityEntityID
        ''' mSecurityEntityProfile.DAL = ConfigSettings.DAL
        ''' mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL)
        ''' mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL)
        ''' mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString
        ''' 
        ''' Dim mBGroups As BGroups = New BGroups(mSecurityEntityProfile, ConfigSettings.CentralManagement)
        ''' ]]>
        ''' </code>
        ''' </example>
        Public Sub New(ByVal securityEntityProfile As MSecurityEntityProfile, ByVal centralManagement As Boolean)
            If securityEntityProfile Is Nothing Then Throw New ArgumentNullException("securityEntityProfile", "securityEntityProfile cannot be a null reference (Nothing in Visual Basic)!")
            If Not centralManagement Then
                If m_DGroups Is Nothing Then
                    m_DGroups = ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DGroups")
                End If
            Else
                m_DGroups = ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DGroups")
            End If
            m_DGroups.ConnectionString = securityEntityProfile.ConnectionString
            m_DGroups.SecurityEntitySeqId = securityEntityProfile.Id
        End Sub

        Public Function GetGroupsBySecurityEntity(ByVal securityEntityId As Integer) As DataTable
            Dim myProfile As New MGroupProfile
            myProfile.SecurityEntityId = securityEntityId
            m_DGroups.Profile = myProfile
            Return m_DGroups.GroupsBySecurityEntity
        End Function

        Public Sub AddGroup(ByVal profile As MGroupProfile)
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!")
            m_DGroups.Profile = profile
            If DatabaseIsOnline() Then m_DGroups.Save()
        End Sub

        Public Function GetProfile(ByVal groupId As Integer) As MGroupProfile
            Dim retProfile As New MGroupProfile
            retProfile.Id = groupId
            m_DGroups.Profile = retProfile
            If DatabaseIsOnline() Then retProfile = New MGroupProfile(m_DGroups.ProfileData)
            Return retProfile
        End Function

        Public Sub DeleteGroup(ByVal profile As MGroupProfile)
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!")
            m_DGroups.Profile = profile
            If DatabaseIsOnline() Then m_DGroups.DeleteGroup()
        End Sub

        Public Sub UpdateGroup(ByVal profile As MGroupProfile)
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!")
            m_DGroups.Profile = profile
            If DatabaseIsOnline() Then m_DGroups.Save()
        End Sub

        ''' <summary>
        ''' Searches the specified search critera.
        ''' </summary>
        ''' <param name="searchCriteria">The search critera.</param>
        ''' <returns>DataTable.</returns>
        Public Function Search(ByVal searchCriteria As MSearchCriteria) As DataTable
            If searchCriteria Is Nothing Then Throw New ArgumentNullException("searchCriteria", "searchCriteria cannot be a null reference (Nothing in Visual Basic)!!")
            Dim mRetVal As DataTable = Nothing
            If DatabaseIsOnline() Then mRetVal = m_DGroups.Search(searchCriteria)
            Return mRetVal
        End Function

        ''' <summary>
        ''' Gets the selected roles.
        ''' </summary>
        ''' <param name="profile">The profile.</param>
        ''' <returns>System.String[][].</returns>
        Public Function GetSelectedRoles(ByVal profile As MGroupRoles) As String()
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!")
            Dim ClientRoles As New ArrayList
            m_DGroups.GroupRolesProfile = profile
            Dim myDataTable As DataTable = Nothing
            If DatabaseIsOnline() Then
                Try
                    myDataTable = m_DGroups.GroupRoles
                    Dim myDR As DataRow
                    For Each myDR In myDataTable.Rows
                        ClientRoles.Add(myDR("Role").ToString)
                    Next
                Catch ex As DataAccessLayerException
                    Throw
                Finally
                    If Not myDataTable Is Nothing Then myDataTable.Dispose()
                End Try
            End If
            Return CType(ClientRoles.ToArray(GetType(String)), String())
        End Function

        ''' <summary>
        ''' Updates the group roles.
        ''' </summary>
        ''' <param name="profile">The profile.</param>
        Public Sub UpdateGroupRoles(ByVal profile As MGroupRoles)
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!!")
            m_DGroups.GroupRolesProfile = profile
            If DatabaseIsOnline() Then m_DGroups.UpdateGroupRoles()
        End Sub
    End Class
End Namespace

